### PR TITLE
AB#2992 -- Demographics gender 

### PR DESCRIPTION
**Pull request by olivierroyESDC**
### Description
Added the gender question and completed part2 of the demographics questionnaire. The form should save the data and send the user to the review page.
### Related Azure Boards Work Items
[AB#2992](https://dev.azure.com/DTS-STN/1fc40a8f-28cf-47bc-b6e4-1c234bd06177/_workitems/edit/2992)

### Checklist

- [x] I have tested the changes locally
- [x] I have updated the documentation if necessary
- [x] I have added/updated tests that prove my fix is effective or that my feature works
- [ x] I have checked that my code follows the project's coding style by running `npm run format:check`
- [x] I have checked that my code contains no linting errors by running `npm run lint`
- [x] I have checked that my code contains no type errors by running `npm run typecheck`
- [ ] I have checked that all unit tests pass by running `npm run test:unit -- run`
- [ ] I have checked that all e2e tests pass by running `npm run test:e2e`

### Test Instructions
For now, go to /apply, start a new questionnaire and manually replace the end of the url with /demographics-part2

### CODE REVIEW

File: frontend/app/routes/_public+/apply+/$id+/demographics-part2.tsx [hunk 1]
```diff
@@ -19,81 +25,187 @@ export const handle = {
...
 }
 
-export async function action({ request }: ActionFunctionArgs) {
-  return json('');
+export async function action({ request, params }: ActionFunctionArgs) {
+  const applyFlow = getApplyFlow();
+  const { id } = await applyFlow.loadState({ request, params });
+
+  const formSchema = z.object({
+    genderTypeChoosen: z.string({ required_error: 'empty-radio' }),
```
Comment (nicholas-ly) [inline frontend/app/routes/_public+/apply+/$id+/demographics-part2.tsx:50]:
```suggestion
    genderTypeChosen: z.string({ required_error: 'empty-radio' }),
```

Or if we remove the (unnecessary IMO) suffix:
```suggestion
    gender: z.string({ required_error: 'empty-radio' }),
```

File: frontend/app/routes/_public+/apply+/$id+/demographics-part2.tsx [hunk 2]
```diff
@@ -19,81 +25,187 @@ export const handle = {
...
+  const actionData = useActionData<typeof action>();
+  const errorSummaryId = 'error-summary';
+
+  const otherGenderHandler = () => {
+    setOtherGenderChecked(true);
+  };
+
+  const nonOtherGenderHandler = () => {
+    setOtherGenderChecked(false);
+  };
```
Comment (nicholas-ly) [inline frontend/app/routes/_public+/apply+/$id+/demographics-part2.tsx:89]: For consistency, consider not mixing arrow functions with normal functions unless absolutely necessary

```suggestion
  function otherGenderHandler() {
    setOtherGenderChecked(true);
  };

  function nonOtherGenderHandler() {
    setOtherGenderChecked(false);
  };
```

File: frontend/app/routes-flow/apply-flow.ts [hunk 3]
```diff
@@ -62,6 +62,17 @@ const communicationPreferencesStateSchema = z.object({
   preferredLanguage: z.string().min(1),
 });
 
+/**
+ * Schema for communication reference.
+ */
+const demographicsPart2StateSchema = z.object({
+  genderTypeChoosen: z.string().min(1),
```
[the remark this answers is not in the document]
Comment (olivierroyESDC) [reply frontend/app/routes-flow/apply-flow.ts]: Donce

File: frontend/app/routes-flow/apply-flow.ts [hunk 4]
```diff
@@ -62,6 +62,17 @@ const communicationPreferencesStateSchema = z.object({
...
+  genderInputField: z.string().min(1).optional(),
```
[the remark this answers is not in the document]
Comment (olivierroyESDC) [reply frontend/app/routes-flow/apply-flow.ts]: Renamed as requested

File: frontend/app/routes/_public+/apply+/$id+/demographics-part2.tsx [hunk 5]
```diff
@@ -19,81 +25,187 @@ export const handle = {
...
   const sexAtBirthTypes = await getLookupService().getAllSexAtBirthTypes();
   const mouthPainTypes = await getLookupService().getAllMouthPaintTypes();
   const lastTimeDentistVisitTypes = await getLookupService().getAllLastTimeDentistVisitTypes();
   const avoidedDentalCostTypes = await getLookupService().getAllAvoidedDentalCostTypes();
+  const genderTypes = await getLookupService().getAllGenderTypes();
   const applyFlow = getApplyFlow();
   const { id, state } = await applyFlow.loadState({ request, params });
-  return json({ sexAtBirthTypes, mouthPainTypes, lastTimeDentistVisitTypes, avoidedDentalCostTypes, id, state });
+
+  const otherGenderCode = genderTypes.find((method) => method.id === OTHER_GENDER_TYPE_ID);
```
[the remark this answers is not in the document]
Comment (olivierroyESDC) [reply frontend/app/routes/_public+/apply+/$id+/demographics-part2.tsx:37]: arrows removed

File: frontend/app/routes/_public+/apply+/$id+/demographics-part2.tsx (hunk 1 above)
Comment (olivierroyESDC) [reply frontend/app/routes/_public+/apply+/$id+/demographics-part2.tsx]: Renamed as requested

File: frontend/app/routes/_public+/apply+/$id+/demographics-part2.tsx (hunk 2 above)
Comment (olivierroyESDC) [reply frontend/app/routes/_public+/apply+/$id+/demographics-part2.tsx]: arrows removed

File: frontend/app/routes/_public+/apply+/$id+/demographics-part2.tsx [hunk 6]
```diff
@@ -19,81 +25,187 @@ export const handle = {
...
+
+  const errorSummaryItems = createErrorSummaryItems(errorMessages);
+  useEffect(() => {
+    if (actionData?.formData && hasErrors(actionData.formData)) {
+      scrollAndFocusToErrorSummary(errorSummaryId);
+    }
+  }, [actionData]);
+
+  const nonOtherGenderTypeOptions: InputRadiosProps['options'] = genderTypes
+    .filter((method) => method.id !== otherGenderCode.id)
```
[the remark this answers is not in the document]
Comment (olivierroyESDC) [reply frontend/app/routes/_public+/apply+/$id+/demographics-part2.tsx]: Renamed as requested

File: frontend/app/routes/_public+/apply+/$id+/demographics-part2.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (nicholas-ly) [reply frontend/app/routes/_public+/apply+/$id+/demographics-part2.tsx:37]: This particular comment is just to rename `method` (which still exists) to `genderType`.

File: frontend/app/routes/_public+/apply+/$id+/demographics-part2.tsx [hunk 7]
```diff
@@ -19,81 +25,178 @@ export const handle = {
...
+    }
+  }, [actionData]);
+
+  const nonOtherGenderTypeOptions: InputRadiosProps['options'] = genderTypes
+    .filter((genderType) => genderType.id !== otherGenderCode.id)
+    .map((method) => ({
+      children: i18n.language === 'fr' ? method.nameFr : method.nameEn,
+      value: method.id,
+      //defaultChecked: state && state.gender !== otherGenderCode.id,
+      onClick: nonOtherGenderHandler,
```
[the remark this answers is not in the document]
Comment (nicholas-ly) [reply frontend/app/routes/_public+/apply+/$id+/demographics-part2.tsx:128]: We can and should also use the `getNameByLanguage(..)` utility function:

```suggestion
    .map((genderType) => ({
      children: getNameByLanguage(i18n.language, genderType)
      value: genderType.id,
      onClick: nonOtherGenderHandler,
```

[the remark this answers is not in the document]
Comment (olivierroyESDC) [reply frontend/app/routes/_public+/apply+/$id+/demographics-part2.tsx]: Done

File: frontend/app/routes/_public+/apply+/$id+/demographics-part2.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (olivierroyESDC) [reply frontend/app/routes/_public+/apply+/$id+/demographics-part2.tsx]: i pushed a change, it should be good now
thanks

**Review by nicholas-ly — APPROVED** [review]

**Pull request merged**